Fix double free in criteria

RyanDwyer opened the pull request:
This restores the criteria implementation (after SirCmpwn's revert) and fixes the issue described in #1966. The fix is in a separate commit from the reverts for easy reviewing.

The config used in that issue uses string values for `id` which triggers errors. The caller frees the error, and when the criteria is parsed a second time it attempted to free the error again, causing the crash.

emersion commented on sway/criteria.c:
```diff
@@ -417,10 +417,9 @@ static void unescape(char *value) {
  */
 struct criteria *criteria_parse(char *raw, char **error_arg) {
-	free(error);
 	error = NULL;
```
Maybe set `*error_arg` to NULL here, just to prevent bad surprises?

SirCmpwn commented:
Does this also introduce https://github.com/swaywm/sway/issues/1965 ?

RyanDwyer commented:
It does now.

I had somehow reached the conclusion that `seat_get_focus(seat)` was the same thing as `seat_get_focus_inactive(seat, &root_container)` and as part of the criteria overhaul, replaced it with `seat_get_focus`. I've since learned it's not the same thing. `seat_get_focus(seat)` does an additional check on the `seat->has_focus` property. So ignoring that check meant that `seat_get_focus` can rightfully return `NULL`.

I've changed it back to `seat_get_focus_inactive(seat, &root_container)` which is what the current master branch does and it works fine now.

emersion commented:
Can you provide a test plan for this PR?

RyanDwyer commented:
Sure.

Test sway won't crash on launch when criteria contains multiple errors by launching with the following config (proves #1966 is fixed):

    for_window [id="nautilus"] border pixel
    for_window [id="gnome-calculator"] floating enable; border pixel
    for_window [class="Chromium"] border none
    for_window [title=".*QEMU.*"] border none; fullscreen enable
    for_window [class="qutebrowser"] border none

Test sway doesn't crash when launching programs from dmenu (proves #1965 and #1963 are fixed):

* No special config needed, just launch a program from dmenu.

As well as the test plans in the original PR (#1943).
